query-coordinator/src/test/scala/com/socrata/querycoordinator/QueryRewritingTestUtility.scala

```diff
@@ -0,0 +1,120 @@
+package com.socrata.querycoordinator
+
+import com.socrata.querycoordinator.QueryRewriter.{ColumnId, RollupName}
+import com.socrata.soql._
+import com.socrata.soql.ast.Select
+import com.socrata.soql.collection.OrderedMap
+import com.socrata.soql.environment.{ColumnName, DatasetContext}
+import com.socrata.soql.typed.CoreExpr
+import com.socrata.soql.types.{SoQLType, SoQLValue}
+import org.scalatest.Matchers.{convertToAnyShouldWrapper, equal}
+
+object QueryRewritingTestUtility {
+  // Some type definitions for easier use
+  type DatasetDefinitions = Map[String, Map[String, (SoQLType, String)]]
+  type AnalyzedDatasetContext = AnalysisContext[SoQLType, SoQLValue]
+  type ColumnMappings = Map[QualifiedColumnName, String]
+  type ParsedSoql = BinaryTree[Select]
+  type AnalyzedSoql = BinaryTree[SoQLAnalysis[ColumnName, SoQLType]]
+  type RemappedAnalyzedSoql = BinaryTree[SoQLAnalysis[ColumnId, SoQLType]]
+  type SoqlAnalysisFunction = AnalyzedDatasetContext => ParsedSoql => AnalyzedSoql
+  type RemapAnalyzedSoqlFunction = ColumnMappings => AnalyzedSoql => RemappedAnalyzedSoql
+  type AnalyzeSoqlAndRemapFunction = ParsedSoql => RemappedAnalyzedSoql
+  type SoqlParseFunction = String => BinaryTree[Select]
+  type RollupsDefinition = Map[String, String]
+  type Rewrites = (RemappedAnalyzedSoql, Seq[String])
+  type RemappedRollupAnalysis = Map[RollupName, RemappedAnalyzedSoql]
+  type QueryDefinition = String
+  type RewriteFunction = (RemappedAnalyzedSoql, RemappedRollupAnalysis) => Rewrites
+  // Crazy function that helps build what the rewritten query should be
+  type BuildRewrittenContextFunction = (RollupsDefinition, SoqlParseFunction, AnalyzedDatasetContext, SoqlAnalysisFunction, ColumnMappings, RemapAnalyzedSoqlFunction) => (RemappedAnalyzedSoql, Seq[String])
+
+
+  // This is the test method to use to assert query rewriting functionality
+  def AssertRewrite(soqlParseFunction: SoqlParseFunction, analysisFunction: SoqlAnalysisFunction, analysisMappingFunction: RemapAnalyzedSoqlFunction, rewriteFunction: RewriteFunction)(datasetDefinitions: DatasetDefinitions, rollupsDefinition: RollupsDefinition, queryDefinition: QueryDefinition, expected: BuildRewrittenContextFunction): Unit = {
+    val (datasetContext, columnMappings) = buildDatasetContextAndMapping(datasetDefinitions)
+    val rollupAnalysis = buildRollups((a) => analysisMappingFunction(columnMappings)(analysisFunction(datasetContext)(a)), soqlParseFunction)(rollupsDefinition)
+    val query = analysisMappingFunction(columnMappings)(analysisFunction(datasetContext)(soqlParseFunction(queryDefinition)))
+    val actual: Rewrites = rewriteFunction(query, rollupAnalysis)
+    actual should equal(expected(rollupsDefinition, soqlParseFunction, datasetContext, analysisFunction, columnMappings, analysisMappingFunction))
+
+  }
+
+  def buildRollups(anal: AnalyzeSoqlAndRemapFunction, parser: SoqlParseFunction)(rollups: RollupsDefinition): RemappedRollupAnalysis = {
```

Comment:
is there a use-case for using methods like this and buildDatasetcontext, etc outside of this module? If not, maybe consider making these private.

Reply:
Nope actually I intended for those to be private but missed the modifier, ill add that now